### DISCUSSION TITLE
Error received for non-pending request_type 48 and request ID 441

### DISCUSSION THREAD

**Discussion by vaibhav-rbs** [crossbario/autobahn-python#1649] [Q&A]
Hi,

I am trying to use Autobahn python client (autobahn==20.7.1) based application, the application works fine for most of the times, but sometimes some error occurs which causes the application to terminate itself. I am trying to find the root cause of this issue, also I am trying to see how I can avoid restarting of the application. Many thanks.

full traceback

```
CRITICAL:autobahn.asyncio.websocket.WebSocketClientProtocol:Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/autobahn/wamp/websocket.py", line 95, in onMessage
    self._session.onMessage(msg)
  File "/usr/local/lib/python3.6/site-packages/autobahn/wamp/protocol.py", line 1236, in onMessage
    raise ProtocolError("WampAppSession.onMessage(): ERROR received for non-pending request_type {0} and request ID {1}".format(msg.request_type, msg.request))
autobahn.wamp.exception.ProtocolError: WampAppSession.onMessage(): ERROR received for non-pending request_type 48 and request ID 441

WARNING:autobahn.asyncio.websocket.WebSocketClientProtocol:failing WebSocket connection (code=1002): "WAMP Protocol Error (WampAppSession.onMessage(): ERROR received for non-pending request_type 48 and request ID 441)"
WARNING:autobahn.asyncio.websocket.WebSocketClientProtocol:dropping connection to peer tcp4:172.20.90.237:8080 with abort=True: I dropped the WebSocket TCP connection: WAMP Protocol Error (WampAppSession.onMessage(): ERROR received for non-pending request_type 48 and request ID 441)
INFO:autobahn.asyncio.component.Component:session leaving 'wamp.close.transport_lost'
INFO:app:Session disconnected
INFO:app:Exiting after session disconnect
```

**Comment by oberstet** [comment]
Which WAMP router? This seems buggy. Did you try running your WAMP client against Crossbar.io?

**Comment by vaibhav-rbs** [comment]
yes we we are using Crossbar.io 20.7.1

**Comment by oberstet** [comment]
from the traceback, it seems your autobahn-python based WAMP client bails out from here

https://github.com/crossbario/autobahn-python/blob/7bc85b34e200640ab98a41cfddb38267f39bc92e/autobahn/wamp/protocol.py#L1299

because it received a WAMP ERROR from the router (your Crossbar.io) with a `request_type {0}` which is an invalid `request_type` 

I can't remember an error in Crossbar.io which would fix such behavior, nor do I believe Crossbar.io ever was doing such thing .. so please submit a complete self-contained test including your client and your Crossbar.io version as Docker images which demonstrates the problem. well, if you want anyone to have the necessary prerequisites so that donating free time to have a look makes sense anyways that is;)

**Reply by vaibhav-rbs** [reply]
@oberstet I think you are referring to the string format, however request type in this case is 48 
```
autobahn.wamp.exception.ProtocolError: WampAppSession.onMessage(): ERROR received for non-pending request_type 48 and request ID 441
```

**Reply by oberstet** [reply]
ah, sorry, you are right! forget my complaints for a second then. message type 48 is indeed valid, it is a CALL

https://github.com/crossbario/autobahn-python/blob/7bc85b34e200640ab98a41cfddb38267f39bc92e/autobahn/wamp/message.py#L3945

thus: your client receives an ERROR relating to a CALL it initially sent, but for which is misses data with that request ID 441

I say "misses", but I would guess it once had such data, but it is no longer around in your client.

Such things can happen in distributed systems due to race conditions.

These might be present in WAMP itself (race conditions in WAMP as the protocol level -- I would no bet my life on it .. sorry, we are working on it), or, much more likely, at the application level.

And "the" application means your code;)

Anyways. _If_ that (my theory) is true, _then_ you need the stuff I mentioned above - and more - to even be able to analyze and proof it, not to speak of fix it. Sorry, I know, "bad news":( Don't shoot the messenger ..

-----

> anyways, it's late here and I need to catch some sleep;) cu!
